docs/docs/remark-ast.md

```diff
@@ -0,0 +1,201 @@
+
+As an example from `unist-util-visit`'s readme, it allows for an easy interface to visit particular nodes based on a particular type:
+
+```js
+var remark = require("remark")
+var visit = require("unist-util-visit")
+
+var tree = remark().parse("Some _emphasis_, **importance**, and `code`.")
+
+visit(tree, "text", visitor)
+
+function visitor(node) {
+  console.log(node)
+}
+```
+
+Here, it finds all text nodes and will `console.log` the nodes.
+
+Now with this, you can take the AST from your plugin and get working on adding functionality for headers:
+
+```js:title=plugins/gatsby-remark-purple-headers/index.js
+const visit = require("unist-util-visit")
+
+module.exports = ({ markdownAST }, pluginOptions) => {
+  visit(markdownAST, "heading", node => {
+    // Do stuff with heading nodes
+  })
+
+  return markdownAST
+}
+```
+
+Next, by visiting all heading nodes and passing them into a transformer function, you can manipulate the particular nodes to match your usecase.
+
+Looking again at the AST node for header:
+
+```JSON
+{
+  "type": "heading",
+  "depth": 1,
+  "children": [
+    {
+      "type": "text",
+      "value": "Hello World!",
+      "position": {
+        "start": { "line": 1, "column": 3, "offset": 2 },
+        "end": { "line": 1, "column": 15, "offset": 14 },
+        "indent": []
+      }
+    }
+  ],
+  "position": {
+    "start": { "line": 1, "column": 1, "offset": 0 },
+    "end": { "line": 1, "column": 15, "offset": 14 },
+    "indent": []
+  }
+},
+```
+
+You have context about the text as well as what depth the header is (for instance here you have a depth of 1 which would equate to an `h1` element)
+
+With the inner function of the `visit` call, you parse out all of the text and if it will map to a h1, you set the type of the node to `html` and set the node's value to be some custom HTML.
+
+```js
+const visit = require("unist-util-visit")
+const toString = require("mdast-util-to-string")
+
+module.exports = ({ markdownAST }, pluginOptions) => {
+  visit(markdownAST, "heading", node => {
+    let { depth } = node
+
+    // Skip if not an h1
+    if (depth !== 1) return
+
+    // Grab the innerText of the heading node
+    let text = toString(node)
+
+    const html = `
+        <h1 style="color: rebeccapurple">
+          ${text}
+        </h1>
+      `
+
+    node.type = "html"
+    node.children = undefined
+    node.value = html
+  })
+
+  return markdownAST
+}
+```
+
+A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) which is also made by Unified was used to extract the plaintext of the inner nodes, so this would remove links or other types of nodes inside the header, but given you have full access to the markdown AST, you can modify it however you wish.
```

Comment:
```suggestion
A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) which is provided by Unified was used to extract the plain text of the inner nodes, so this would remove links or other types of nodes inside the header, but given you have full access to the markdown AST, you can modify it however you wish.
```

Also would probably tweak this sentence--it's a bit of a doozy 😩

Reply:
I think it could be even shorter/broken up:
```suggestion
A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) by Unified was used to extract the plain text of the inner nodes. This would remove links or other types of nodes inside the heading, but given you have full access to the markdown AST, you can modify it however you wish.
```